In [220]:
import time
import math

import numpy as np
import matplotlib.pyplot as plt

# Sudoku Solver in Python Using Backtracking Approach

## Introduction
Sudoku is a popular logic puzzle that is played on a 9x9 grid. The goal is to fill all empty cells with numbers from 1 to 9 ensuring that each number appears only once in every row, column, and 3x3 subgrid.

I chose Sudoku for this project because it is easy to understand as a game but not so easy to solve algorithmically. At first I was not sure which topic to choose and I was considering other options such as Dobble. Sudoku seemed like a good balance between being understandable and still challenging. Also it is a good example of how mathematical rules can be translated into code.

The solution uses a backtracking approach because it is relatively easy to  implement, guarantees finding a solution and provides a base for optimizations through other techniques such as constraint propagation, Dancing Links and SAT solvers. 

## Mathematical Formulation
Sudoku can be viewed as a **constraint satisfaction problem (CSP)**. The board is represented as a $9 \times 9$ matrix where $a_{ij}=0$ indicates an empty cell.
Each cell must be filled with a value from the set $D = \{1, 2, ..., 9\}$

The number must be unique in its corresponding:
1. **row:** Each number $n$ appears exactly once in every row $i$:
$$\forall i, n: \sum_{j=1}^{9} [a_{ij} = n] = 1$$
2. **column:** Each number $n$ appears exactly once in every column $j$:
$$\forall j, n: \sum_{i=1}^{9} [a_{ij} = n] = 1$$
3. **subgrid:** Each number $n$ appears exactly once in every $3 \times 3$ subgrid:
$$\forall r, c \in \{0,1,2\}, n: \sum_{i=0}^{2} \sum_{j=0}^{2} [a_{(3r+i+1)(3c+j+1)} = n] = 1$$

Therefore solving Sudoku means finding values for all empty cells while following these constraints simultaneously.


## What is Backtracking
Backtracking is a recursive search algorithm that explores possible solutions step by step. It works by:
1. finding an empty cell
2. assigning a possible value from 1 to 9
3. checking if a value is valid
4. proceeding recursively if the value is valid
5. going back and trying another value if the assignment violates the constraints

This makes backtracking an effective method for solving Sudoku as the puzzle relies on rules that define valid and invalid choices.

## Example Sudoku Grid

This Sudoku example represents a medium to hard difficulty level, as it contains few initial values and requires deeper recursive exploration to reach a solution.

In [221]:
grid = [
    [0, 0, 0, 0, 0, 0, 2, 0, 0],
    [0, 8, 0, 0, 0, 7, 0, 9, 0],
    [6, 0, 2, 0, 0, 0, 5, 0, 0],
    [0, 7, 0, 0, 6, 0, 0, 0, 0],
    [0, 0, 0, 9, 0, 1, 0, 0, 0],
    [0, 0, 0, 0, 2, 0, 0, 4, 0],
    [0, 0, 5, 0, 0, 0, 6, 0, 3],
    [0, 9, 0, 4, 0, 0, 0, 7, 0],
    [0, 0, 6, 0, 0, 0, 0, 0, 0]
]

## Print the Sudoku Grid

In [222]:
def print_grid(grid):
    for row_index in range(len(grid)):
        if row_index % 3 == 0 and row_index != 0:
            print("-" * 21)
        
        for col_index in range(len(grid[0])):
            if col_index % 3 == 0 and col_index != 0:
                print("|", end=" ")
            
            if col_index == 8:
                print(grid[row_index][col_index])
            else:
                print(grid[row_index][col_index], end=" ")

## Find an Empty Cell
The algorithm first needs to find and empty cell. In the implementation below 0 means that the cell is empty.


In [223]:
def find_empty(grid):
    for row_index in range(9):
        for col_index in range(9):
            if grid[row_index][col_index] == 0:
                return(row_index, col_index)
    return None

## Check if a Value is Valid
Before placing a number, we must check whether it satisfies the Sudoku constraints.

Thе function below checks whether a number can be placed in a given position by verifying the row, column and 3x3 subgrid constraints.

In [224]:
def is_valid(grid, number, position):
    row_index = position[0]
    col_index = position[1]

    for col in range(9):
        if grid[row_index][col] == number:
            if col != col_index:
                return False

    for row in range(9):
        if grid[row][col_index] == number:
            if row != row_index:
                return False

    box_row = row_index // 3
    box_col = col_index // 3

    for row in range(box_row * 3, box_row * 3 + 3):
        for col in range(box_col * 3, box_col * 3 + 3):
            if grid[row][col] == number:
                if (row, col) != position:
                    return False

    return True

## Backtracking Solver
1. The solver tries numbers from 1 to 9 in each empty cell
2. If a number is valid, the algorithm proceeds to the next empty cell
3. If no valid number leads to a solution, the algorithm returns to the previous step and tests another value

In [225]:
def solve(grid):
    empty_position = find_empty(grid)

    if empty_position is None:
        return True

    row_index = empty_position[0]
    col_index = empty_position[1]

    for number in range(1, 10):
        if is_valid(grid, number, (row_index, col_index)):
            grid[row_index][col_index] = number

            if solve(grid):
                return True

            grid[row_index][col_index] = 0

    return False

## Solve the Sudoku

In [226]:
print("Sudoku grid before solving:")
print_grid(grid)

print("\nSolving the Sudoku...\n")

if solve(grid):
    print("Solved Sudoku grid:")
    print_grid(grid)
else:
    print("No solution exists.")

Sudoku grid before solving:
0 0 0 | 0 0 0 | 2 0 0
0 8 0 | 0 0 7 | 0 9 0
6 0 2 | 0 0 0 | 5 0 0
---------------------
0 7 0 | 0 6 0 | 0 0 0
0 0 0 | 9 0 1 | 0 0 0
0 0 0 | 0 2 0 | 0 4 0
---------------------
0 0 5 | 0 0 0 | 6 0 3
0 9 0 | 4 0 0 | 0 7 0
0 0 6 | 0 0 0 | 0 0 0

Solving the Sudoku...

Solved Sudoku grid:
9 5 7 | 6 1 3 | 2 8 4
4 8 3 | 2 5 7 | 1 9 6
6 1 2 | 8 4 9 | 5 3 7
---------------------
1 7 8 | 3 6 4 | 9 5 2
5 2 4 | 9 7 1 | 3 6 8
3 6 9 | 5 2 8 | 7 4 1
---------------------
8 4 5 | 7 9 2 | 6 1 3
2 9 1 | 4 3 6 | 8 7 5
7 3 6 | 1 8 5 | 4 2 9


## Testing with Different Difficulty Levels

To analyze the performance of the backtracking algorithm we test it on three puzzles with different difficulty levels. The difficulty is determined by the number of initially filled cells — the fewer the filled cells, the harder the puzzle.

In [227]:
test_grids = [
    ("Easy", [
        [0, 0, 3, 0, 2, 0, 6, 0, 0],
        [9, 0, 0, 3, 0, 5, 0, 0, 1],
        [0, 0, 1, 8, 0, 6, 4, 0, 0],
        [0, 0, 8, 1, 0, 2, 9, 0, 0],
        [7, 0, 0, 0, 0, 0, 0, 0, 8],
        [0, 0, 6, 7, 0, 8, 2, 0, 0],
        [0, 0, 2, 6, 0, 9, 5, 0, 0],
        [8, 0, 0, 2, 0, 3, 0, 0, 9],
        [0, 0, 5, 0, 1, 0, 3, 0, 0]
    ]),
    ("Medium", [
        [0, 0, 0, 8, 0, 0, 0, 0, 0],
        [3, 0, 0, 0, 0, 0, 0, 2, 0],
        [0, 0, 4, 0, 0, 6, 0, 0, 0],
        [0, 0, 0, 0, 4, 0, 7, 0, 0],
        [6, 0, 0, 0, 0, 0, 0, 0, 9],
        [0, 0, 7, 0, 6, 0, 0, 0, 0],
        [0, 0, 0, 4, 0, 0, 9, 0, 0],
        [0, 5, 0, 0, 0, 0, 0, 0, 6],
        [0, 0, 0, 0, 0, 5, 0, 0, 0]
    ]),
    ("Hard", [
        [0, 0, 1, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 5, 0, 0, 0, 0, 4],
        [0, 0, 0, 0, 0, 0, 0, 3, 0],
        [0, 0, 0, 0, 3, 0, 0, 0, 0],
        [0, 4, 0, 0, 0, 0, 2, 0, 0],
        [0, 0, 0, 0, 0, 6, 0, 0, 0],
        [0, 7, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 8, 0, 0, 0],
        [0, 0, 0, 0, 1, 0, 5, 0, 0]
    ])
]

In [228]:
col1_width = len("Difficulty")
col2_width = len("Known cells")

column_names = f"{'Difficulty':<{col1_width}} {'Known cells':<{col2_width}} {'Solution'}"
line = "-" * (col1_width + col2_width + len("Solution") + 2)
print(column_names)
print(line)

for name, grid in test_grids:

    known_cells = 0
    for row_index in range(9):
        for col_index in range(9):
            if grid[row_index][col_index] != 0:
                known_cells += 1

    is_solved = solve(grid)

    if is_solved:
        solution = "Solved"
    else:
        solution = "No solution"

    print(f"{name:<{col1_width}} {known_cells:<{col2_width}} {solution}")

Difficulty Known cells Solution
-------------------------------
Easy       32          Solved
Medium     16          Solved
Hard       12          Solved


## Count the Backtracking Steps
In orer to further analyze the algorithm we count the number of backtracking steps for each sudoku puzzle. A backtracking step occurs every time the algorithm has a wrong choise and tries a different number.

In [229]:
test_grids = [
    ("Easy", [
        [0, 0, 3, 0, 2, 0, 6, 0, 0],
        [9, 0, 0, 3, 0, 5, 0, 0, 1],
        [0, 0, 1, 8, 0, 6, 4, 0, 0],
        [0, 0, 8, 1, 0, 2, 9, 0, 0],
        [7, 0, 0, 0, 0, 0, 0, 0, 8],
        [0, 0, 6, 7, 0, 8, 2, 0, 0],
        [0, 0, 2, 6, 0, 9, 5, 0, 0],
        [8, 0, 0, 2, 0, 3, 0, 0, 9],
        [0, 0, 5, 0, 1, 0, 3, 0, 0]
    ]),
    ("Medium", [
        [0, 0, 0, 8, 0, 0, 0, 0, 0],
        [3, 0, 0, 0, 0, 0, 0, 2, 0],
        [0, 0, 4, 0, 0, 6, 0, 0, 0],
        [0, 0, 0, 0, 4, 0, 7, 0, 0],
        [6, 0, 0, 0, 0, 0, 0, 0, 9],
        [0, 0, 7, 0, 6, 0, 0, 0, 0],
        [0, 0, 0, 4, 0, 0, 9, 0, 0],
        [0, 5, 0, 0, 0, 0, 0, 0, 6],
        [0, 0, 0, 0, 0, 5, 0, 0, 0]
    ]),
    ("Hard", [
        [0, 0, 1, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 5, 0, 0, 0, 0, 4],
        [0, 0, 0, 0, 0, 0, 0, 3, 0],
        [0, 0, 0, 0, 3, 0, 0, 0, 0],
        [0, 4, 0, 0, 0, 0, 2, 0, 0],
        [0, 0, 0, 0, 0, 6, 0, 0, 0],
        [0, 7, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 8, 0, 0, 0],
        [0, 0, 0, 0, 1, 0, 5, 0, 0]
    ])
]

In [230]:
backtrack_counter = [0]

def solve_count(grid):
    empty_position = find_empty(grid)

    if empty_position is None:
        return True

    row_index = empty_position[0]
    col_index = empty_position[1]

    for number in range(1, 10):
        if is_valid(grid, number, (row_index, col_index)):
            grid[row_index][col_index] = number

            if solve_count(grid):
                return True

            grid[row_index][col_index] = 0
            backtrack_counter[0] += 1

    return False

col1_width = len("Difficulty")
col2_width = len("Known cells")
col3_width = len("Backtrack steps")

column_names = f"{'Difficulty':<{col1_width}} {'Known cells':<{col2_width}} {'Backtrack steps'}"
line = "-" * (col1_width + col2_width + col3_width + 4)
print(column_names)
print(line)

for name, grid in test_grids:

    known_cells = 0
    for row_index in range(9):
        for col_index in range(9):
            if grid[row_index][col_index] != 0:
                known_cells += 1

    backtrack_counter[0] = 0
    solve_count(grid)

    print(f"{name:<{col1_width}} {known_cells:<{col2_width}} {backtrack_counter[0]}")

Difficulty Known cells Backtrack steps
----------------------------------------
Easy       32          151
Medium     16          37
Hard       12          597


## Solving Time
Now lets count the time that it takes to solve each sudoku puzzle. We expect the execution time to increase with difficulty as the algorithm needs to explore more combinations.

In [233]:
test_grids = [
    ("Easy", [
        [0, 0, 3, 0, 2, 0, 6, 0, 0],
        [9, 0, 0, 3, 0, 5, 0, 0, 1],
        [0, 0, 1, 8, 0, 6, 4, 0, 0],
        [0, 0, 8, 1, 0, 2, 9, 0, 0],
        [7, 0, 0, 0, 0, 0, 0, 0, 8],
        [0, 0, 6, 7, 0, 8, 2, 0, 0],
        [0, 0, 2, 6, 0, 9, 5, 0, 0],
        [8, 0, 0, 2, 0, 3, 0, 0, 9],
        [0, 0, 5, 0, 1, 0, 3, 0, 0]
    ]),
    ("Medium", [
        [0, 0, 0, 8, 0, 0, 0, 0, 0],
        [3, 0, 0, 0, 0, 0, 0, 2, 0],
        [0, 0, 4, 0, 0, 6, 0, 0, 0],
        [0, 0, 0, 0, 4, 0, 7, 0, 0],
        [6, 0, 0, 0, 0, 0, 0, 0, 9],
        [0, 0, 7, 0, 6, 0, 0, 0, 0],
        [0, 0, 0, 4, 0, 0, 9, 0, 0],
        [0, 5, 0, 0, 0, 0, 0, 0, 6],
        [0, 0, 0, 0, 0, 5, 0, 0, 0]
    ]),
    ("Hard", [
        [0, 0, 1, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 5, 0, 0, 0, 0, 4],
        [0, 0, 0, 0, 0, 0, 0, 3, 0],
        [0, 0, 0, 0, 3, 0, 0, 0, 0],
        [0, 4, 0, 0, 0, 0, 2, 0, 0],
        [0, 0, 0, 0, 0, 6, 0, 0, 0],
        [0, 7, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 8, 0, 0, 0],
        [0, 0, 0, 0, 1, 0, 5, 0, 0]
    ])
]

col1_width = len("Difficulty")
col2_width = len("Known cells")
col3_width = len("Time(seconds)")

column_names = f"{'Difficulty':<{col1_width}} {'Known cells':<{col2_width}} {'Time(seconds)'}"
line = "-" * (col1_width + col2_width + col3_width + 4)

print(column_names)
print(line)

for name, grid in test_grids:

    known_cells = 0
    for row_index in range(9):
        for col_index in range(9):
            if grid[row_index][col_index] != 0:
                known_cells += 1

    start_time = time.time()
    solve(grid)
    end_time = time.time()

    solving_time = round(end_time - start_time, 3)

    print(f"{name:<{col1_width}} {known_cells:<{col2_width}} {solving_time}")

Difficulty Known cells Time(seconds)
--------------------------------------
Easy       32          0.003
Medium     16          0.001
Hard       12          0.057


## Alternative Approaches for Solving Sudoku

Backtracking is not the only way to solve Sudoku. There are several alternative approaches that are more efficient but also more complex for implementation.

**Constraint Propagation** reduces the number of possible values for each cell before the search begins. Instead of trying all possibilities it checks whether each value is still valid before placing it in a cell. For example if a row already contains 4 that value is automatically removed from all other empty cells in the row. This can be extended with more advanced techniques such as Forward Checking and Arc Consistency (AC-3). These techniques significantly reduce the number of choices the algorithm needs to explore compared to pure backtracking.

**Dancing Links** (introduced by Donald Knuth) solves Sudoku by transforming it into an Exact Cover problem. This is a mathematical problem where the goal is to select a set of rows from a matrix such that every column is covered exactly once. The key innovation is a special data structure that allows very fast addition and removal of elements making backtracking highly efficient. In practice this approach is significantly more efficient than naive backtracking.

**SAT Solvers** convert Sudoku into a Boolean Satisfiability Problem (SAT) where each constraint becomes a logical equation. Specialized algorithms then find values that satisfy all equations simultaneously. Modern SAT solvers are very powerful for large-scale constraint satisfaction problems.

### Complexity Comparison

| Approach | Practical Performance |
|---|---|
| Backtracking | Slow without pruning |
| Constraint Propagation | Moderate improvement |
| Dancing Links | Very fast in practice |
| SAT Solvers | Highly optimized |

### Why Backtracking

I chose backtracking for this project because it is the most straightforward approach and clearly demonstrates the mathematical constraints of Sudoku. While more advanced methods offer better performance backtracking is easier to understand, easier to implement and serves as a solid baseline for further optimization.

### Real-World Relevance

Sudoku is a classic example of a Constraint Satisfaction Problem (CSP). Similar techniques are widely used in task scheduling, query optimization in databases and constraint-based planning systems. All of these are directly relevant to data engineering.

## Complexity and Analysis

The backtracking algorithm is simple and effective but it may become slow for difficult Sudoku puzzles because it explores many possible combinations.

In the worst case the time complexity is very high because the algorithm may need to try many values recursively. However for standard Sudoku puzzles backtracking is often fast enough and provides a clear example of recursive problem solving.

The results from our tests show an interesting observation — the difficulty of a puzzle does not depend only on the number of нхдух cells, but also on their distribution. In our tests, the Medium puzzle had fewer backtracking steps (37) than the Easy puzzle (151) even though it had fewer given cells (16 vs 32). This suggests that the arrangement of the  known cells plays a key role in how efficiently the algorithm finds a solution.

This project shows how a puzzle can be modeled mathematically and solved algorithmically using Python.

## Conclusion

In this project, I represented Sudoku as a constraint satisfaction problem and solved it using backtracking.
The project combines mathematical reasoning, recursion, and Python implementation.
It also shows how a seemingly simple puzzle can lead to a rich algorithmic problem.